File: server/services/mqtt.py
# Filename: mqtt.py
# Description: Establishes a connection with an MQTT Broker.

import paho.mqtt.client as mqtt
from db.repository import write_data

# Callback functions
def on_connect(client, userdata, flags, rc):
    print("Connected to MQTT Broker")
    client.subscribe("lnu/iot/al227bn/sensor")

def on_message(client, userdata, msg):
    db = userdata["db_client"]
    write_data(db, msg.payload)


def create_mqtt_client():
    return mqtt.Client()

def connect(client):
    """Connects to an MQTT broker."""
    client.on_connect = on_connect
    client.on_message = on_message

    client.connect("broker.emqx.io", 1883, 60)
    client.loop_start()

File: client/src/components/mqtt.ts
/**
 * Sets up an MQTT connection over a Web Socket instance.
 */

import mqtt from "mqtt"

export function connectMQTT() {
  const clientID = "mqtt_" + Math.random().toString(16).substring(2, 10)
  const host = "wss://broker.emqx.io:8084/mqtt"
  const options: mqtt.IClientOptions = {
    keepalive: 60,
    clientId: clientID,
    protocolId: "MQTT",
    protocolVersion: 5,
    clean: true,
    reconnectPeriod: 1000,
    connectTimeout: 3 * 1000,
  }

  console.log("Connecting to MQTT Client ...")

  const client = mqtt.connect(host, options)

  client.on("error", (error) => {
    console.error("Connection error: ", error)
    client.end()
  })
  client.on("reconnect", () => {
    console.log("Reconnecting ...")
  })
  client.on("connect", () => {
    console.log(`Client connected: ${clientID}`)
    client.subscribe("lnu/iot/al227bn/sensor", { qos: 0 })
  })
  client.on("message", (topic, message, packet) => {
    console.log(`Received message: ${message.toString()}, on topic: ${topic}`)
  })
}